feat!: add block hash to log store

mverzilli opened the pull request:
We want to expose some event metadata to users of Aztec.js (see: https://github.com/AztecProtocol/aztec-packages/issues/16245).

An important metadata attribute is block hash. Block number is not enough due to the possibliity of reorgs.

Currently, in archiver nodes, logs are stored by block number, and the corresponding block hash is lost. This works at the archiver, because the archiver is designed to prune its stores and start over each time a reorg is detected.

Users who download data from the node, however, have no reasonably efficient way of detecting whether the log data they downloaded is still valid unless they also get a block hash to pin the logs they get to. 

So we can bridge this gap in one of two ways:

1. Whenever the node gets a request for logs, it queries the block store to resolve block hashes from block numbers. This works because archiver provides strong guarantees that the data stored corresponds to only one chain. However, it it introduces more store reads

2. At the time of storing logs, we can also store the block hash corresponding to the block number at the moment of storing. this increments the size needed to store the logs of each indexed block by 1 Fr, but in exchange for that we can just unpack the block hash from the logs we're processing without needing extra reads from block store. It also makes block addition a bit more expensive since now we'll be potentially calculating one hash per block (although since `addBlocks` receives them and `hash()` is cached maybe it's not even the case?) 

This PR implements option 2, but if I'm hitting any no-no's please let me know

Note that this PR bumps the ARCHIVER_DB_VERSION, since it modifies how we index public and contract class logs.   

Closes F-220

spalladino commented on yarn-project/stdlib/src/logs/tx_scoped_l2_log.ts:
```diff
@@ -88,6 +95,7 @@ export class TxScopedL2Log {
       this.dataStartIndexForTx === other.dataStartIndexForTx &&
       this.logIndexInTx === other.logIndexInTx &&
       this.blockNumber === other.blockNumber &&
+      this.blockHash === other.blockHash &&
```
```suggestion
      this.blockHash.equals(other.blockHash) &&
```
We need to set up a linter rule to catch these kind of errors, they happen way too often. That, or convert L2BlockHash into a branded type over a string.

mverzilli replied:
Oof, nice catch

spalladino reviewed (changes requested):
Looks good! Just a small fix that we should have a linter for.

mverzilli committed:
feat!: add block hash to log store

We want to expose some event metadata to users of Aztec.js (see: https://github.com/AztecProtocol/aztec-packages/issues/16245).

An important metadata attribute is block hash. Block number is not enough due to the possibliity of reorgs.

Currently, in archiver nodes, logs are stored by block number, and the corresponding block hash is lost. This works at the archiver, because the archiver is designed to prune its stores and start over each time a reorg is detected.

Users who download data from the node, however, have no reasonably efficient way of detecting whether the log data they downloaded is still valid unless they also get a block hash to pin the logs they get to.

So we can bridge this gap in one of two ways:

1. Whenever the node gets a request for logs, it queries the block store to resolve block hashes from block numbers. This works because archiver provides strong guarantees that the data stored corresponds to only one chain. However, it it introduces more store reads

2. At the time of storing logs, we can also store the block hash corresponding to the block number at the moment of storing. this increments the size needed to store the logs of each indexed block by 1 Fr, but in exchange for that we can just unpack the block hash from the logs we're processing without needing extra reads from block store. It also makes block addition a bit more expensive since now we'll be potentially calculating one hash per block (although since `addBlocks` receives them and `hash()` is cached maybe it's not even the case?)

This PR implements option 2, but if I'm hitting any no-no's please let me know

Note that this PR bumps the ARCHIVER_DB_VERSION, since it modifies how we index public and contract class logs.

Closes F-220
